src/uproot/source/futures.py

```diff
@@ -469,3 +472,44 @@ def __exit__(self, exception_type, exception_value, traceback):
         self.shutdown()
         self._resource.__exit__(exception_type, exception_value, traceback)
         self._closed = True
+
+
+class LoopExecutor:
+    def __repr__(self):
+        return f"<LoopExecutor at 0x{id(self):012x}>"
+
+    def __init__(self):
+        self._loop = asyncio.new_event_loop()
+        self._thread = threading.Thread(target=self._run)
+        self.start()
+
+    def start(self):
+        self._thread.start()
+        return self
+
+    def shutdown(self):
+        self._loop.call_soon_threadsafe(self._loop.stop)
+        self._thread.join()
+
+    def _run(self):
+        asyncio.set_event_loop(self._loop)
+        try:
+            self._loop.run_forever()
+        finally:
+            self._loop.run_until_complete(self._loop.shutdown_asyncgens())
+            self._loop.close()
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.shutdown()
+
+    @property
+    def loop(self) -> asyncio.AbstractEventLoop:
+        return self._loop
+
+    def submit(self, coroutine, *args) -> asyncio.Future:
+        coroutine_object = coroutine(*args)
+        return asyncio.run_coroutine_threadsafe(coroutine_object, self._loop)
```

Comment:
Implementation of the `LoopExecutor`. This is the part that had the possible race condition. I don't know asyncio well enough to be sure that it's resolved, but CI was sensitive to it, so when everything passes, we'll take that as a sign that it's likely okay.